contracts/strategies/examples/assurance-contract/AssuranceContract.sol

```diff
@@ -0,0 +1,130 @@
+            campaign.totalPledged += msg.value;
+            pledges[_poolId][msg.sender] += msg.value;
+        } else {
+            IERC20 token = IERC20(campaign.tokenAddress);
+            uint256 balanceBefore = token.balanceOf(address(this));
+            token.safeTransferFrom(msg.sender, address(this), _amount);
+            uint256 balanceAfter = token.balanceOf(address(this));
+            uint256 actualAmount = balanceAfter - balanceBefore;
+            campaign.totalPledged += actualAmount;
+            pledges[_poolId][msg.sender] += actualAmount;
+        }
+
+        emit Pledged(_poolId, msg.sender, _amount);
+
+        if (campaign.totalPledged >= campaign.goal) {
+            emit GoalReached(_poolId);
+        }
+    }
+
+    function claimFunds(uint256 _poolId) external {
+        Campaign storage campaign = campaigns[_poolId];
+
+        require(block.timestamp >= campaign.deadline, "Campaign not ended");
+        require(campaign.totalPledged >= campaign.goal, "Goal not reached");
+        require(!campaign.finalized, "Funds already claimed");
+
+        campaign.finalized = true;
+        uint256 amount = campaign.totalPledged;
+
+        if (campaign.tokenAddress == address(0)) {
+            (bool success, ) = campaign.beneficiary.call{value: amount}("");
+            require(success, "ETH transfer failed");
+        } else {
+            IERC20 token = IERC20(campaign.tokenAddress);
+            token.safeTransfer(campaign.beneficiary, amount);
+        }
+
+        emit FundsClaimed(_poolId, campaign.beneficiary, amount);
+    }
+
+    function refund(uint256 _poolId) external {
+        Campaign storage campaign = campaigns[_poolId];
+
+        require(block.timestamp >= campaign.deadline, "Campaign not ended");
+        require(campaign.totalPledged < campaign.goal, "Goal was reached");
+        require(!campaign.finalized, "Campaign already finalized");
+
+        uint256 amount = pledges[_poolId][msg.sender];
+        require(amount > 0, "No funds to refund");
+
+        pledges[_poolId][msg.sender] = 0;
+
+        if (campaign.tokenAddress == address(0)) {
+            (bool success, ) = msg.sender.call{value: amount}("");
+            require(success, "ETH transfer failed");
+        } else {
+            IERC20 token = IERC20(campaign.tokenAddress);
+            token.safeTransfer(msg.sender, amount);
+        }
+
+        emit FundsRefunded(_poolId, msg.sender, amount);
+    }
+
+    function _allocate(address[] memory, uint256[] memory, bytes memory, address) internal virtual override {
+        revert("AssuranceContract: Allocate not implemented");
+    }
+
+    function _distribute(address[] memory, bytes memory, address) internal virtual override {
+        revert("AssuranceContract: Distribute not implemented");
+    }
+
+    function _register(address[] memory, bytes memory, address) internal virtual override returns (address[] memory) {
+        revert("AssuranceContract: Register not implemented");
```

Comment:
i think the current allo pool uses just naked addresses.  need to use allo registry items per https://github.com/allo-protocol/allo-v2.1/blob/dev/contracts/core/Allo.md#register-recipient.

would be nice if this could be inherited from an upstream contract..

Reply:
You could make use of the `RecipientsExtension`. Using the `_register` from the extensions includes this check https://github.com/allo-protocol/allo-v2.1/blob/c583e0eda023e75b4cd5519ee011693a229bde75/contracts/strategies/extensions/register/RecipientsExtension.sol#L293-L296